packages/sitecore-jss/src/personalization/layout-personalization-service.ts

```diff
@@ -0,0 +1,131 @@
+    }
+  }
+
+  getPersonalizedComponent(componentUid: string): ComponentRendering | null {
+    return this.personalizedComponents && this.personalizedComponents[componentUid];
+  }
+
+  isLoading() {
+    return this.personalizedComponents == null;
+  }
+
+  loadPersonalizedComponent(componentUid: string): Promise<ComponentRendering | null> {
+    if (this.personalizationResult == null) {
+      return Promise.reject("loadPersonalization should be called before getting personalized component");
+    } else {
+      return new Promise<ComponentRendering | null>((resolve, reject) => {
+        this.personalizationResult?.then(pr => {
+          resolve(pr[componentUid]);
+        }).catch(error => { reject(error) });
+      });
+    }
+  }
+
+  async personalize(
+    context: LayoutServiceContext,
+    personalizedRenderings: PersonalizedComponentRendering[]
+  ): Promise<{ [key: string]: ComponentRendering | null }> {
+    if (personalizedRenderings.length === 0) {
+      return {};
+    }
+    const personalizedRenderingIds = personalizedRenderings.map(r => r.uid);
+    var personalizedFragments : { [key: string]: ComponentRendering | null | undefined } = {};
+    try {
+      var personalizationDecisionsResult = await this.personalizationDecisionsService.getPersonalizationDecisions(context.itemPath as string, context.language as string, personalizedRenderingIds);
+      var renderingsDecisions = personalizationDecisionsResult.renderings;
+      var personalizedFragmentsRequests: Promise<void | LayoutFragmentData>[] = [];
+      Object.keys(renderingsDecisions).map(renderingId => {
+        var variantKey = renderingsDecisions && renderingsDecisions[renderingId]?.variantKey;
+        if (variantKey) {
+          // load fragments in parallel
+          personalizedFragmentsRequests.push(
+            this.layoutFragmentService.fetchLayoutFragmentData(
+              context.itemPath as string,
+              context.language as string,
+              renderingId,
+              variantKey).then(fr => {
+                personalizedFragments[renderingId] = fr.fragment;
+            }).catch((error) => {
+              console.error(error);
+              personalizedFragments[renderingId] = undefined; // default will be used for undefined fragments
+            })
+          );
+        } else if (variantKey === null) { // hidden by personalization
+          personalizedFragments[renderingId] = null;
+        } else {
+          personalizedFragments[renderingId] = undefined; // was not able to resolve decisions for this rendering, default will be used
+        }
+      });
+      // wait all fragments is requisted, no fail on error, default should be applied
+      // Promise.allSettled simple polifil.
+      await Promise.all(personalizedFragmentsRequests.map(p => p.then(() => {}, () => {})));
+    } catch (error) {
+      console.error(error);
+      personalizedRenderingIds.forEach(id => personalizedFragments[id] = undefined );
```

Comment:
The `personalize` function has 3 places where `undefined` value is set.
You can pre-initialize `personalizedFragments` object with `undefined` values once to simplify the code.

Reply:
Unfortunately, an attempt to resolve this led to even more complicated code. We improved comments a bit and going to leave this code as it is.